Update DependencyInstaller.sh to support MacOS

rovinski opened the pull request:
Draft PR to support MacOS. Closes #2173.

This isn't a perfect solution to the problem, but I think it's a good start. There are several issues I worked around while making this.

* Problem: Homebrew's `spdlog` does not work out of the box as of right now.
* Cause: This has been mentioned a few times in issues such as #2173, #2220, and #2097, but the main issue seems to be that `spdlog` is dependent on `fmt`. While `fmt 8` works with OpenROAD's code, `fmt 9` has more strict standards for conversions of objects into strings for logging, thus resulting in several compiler errors.
* Solution: I think that longer term, we simply have to fix all the object to string conversions so that we can upgrade to `spdlog 1.10` which depends on `fmt 9`. This is what Homebrew wants to do, and it makes installing old package versions as difficult as possible. This PR works around this by forcibly installing `fmt 8.1.1_1` and `spdlog 1.9.2`, which I have found to work.

---

* Problem: Some dependencies don't take precedence in the `PATH` over system-installed libraries, which can cause compilation errors (such as bison)
* ~~Cause: `/usr/bin` tends to be in front of `/usr/local`~~ Homebrew refuses to link libs which are already provided by MacOS
* Solution: For right now, I just have the script print commands for the user to copy+paste to update the `PATH`. This is a dependency script, so it's not expected to build `openroad` for the user.

---

* Problem: Xcode Command Line Tools and the Xcode app seem to conflict on my system when both are installed.
* Cause: Not sure. I have consistently found that openroad has std lib reference issues *if* the Command Line Tools are installed (e.g. `xcode-select --install`), regardless of whether the Xcode app is installed. I can only compile if *only* the Xcode app is installed. I have checked `CMakeCache.txt` and cmake appears to be pointing to all the right directories (no mixing between the CLT and the app), so I don't know why this fails. Annoyingly, homebrew refuses to install Python if it doesn't detect CLT (even if the app is installed).
* My current workaround is to install CLT to install the dependencies, then remove CLT and install the Xcode app to actually build openroad. I sincerely hope this is just a problem on my side and not typical.

The script was tested on an x86_64 macbook running `Monterey 12.5.1` with `Apple clang version 13.1.6 (clang-1316.0.21.2.5)`

TODO items
- [x] Separate packages into runtime and development
- [x] See if someone can replicate my issues
- [ ] Add package(s) to enable GUI

maliberty commented:
is this ready for review?

rovinski commented:
Not quite, I'm hoping to get someone else with MacOS to test it out to make sure it works as expected. Also I need to address the TODO items.

If you have feedback, you are welcome to comment though.

maliberty commented:
I checked out spdlog 1.10.0 on my usual linux machine and OR builds without error.  Are you using an external fmt library?

rovinski commented:
Yes, Homebrew packages `fmt` separately from `spdlog`. There isn't a good way to get `spdlog` to build with the bundled `fmt` unless it's built from source.

maliberty commented:
I was able to reproduce it.  I find the changes to fmt to be horrible...

rovinski commented:
I was able to confirm that at least 1 other person with a macbook was able to build using the script.

Maybe I can look into `FMT_DEPRECATED_OSTREAM` from https://github.com/fmtlib/fmt/issues/3081#issuecomment-1238789389 and resolve the spdlog version issue in a separate PR.

maliberty commented:
lgtm, I'll leave it to @vvbandeira to approve.

stefanottili commented:
Having the right fmt/spdlog combination installed with HomeBrew on a MacBook Air M1 Darwing gets OpenRoad to compile again, thanks.

vvbandeira commented on etc/DependencyInstaller.sh:
```diff
@@ -200,6 +200,59 @@ _installCentosRuntime() {
     yum update -y
 }
 
+_installDarwinCleanUp() {
+    echo "";
+}
+
+_installDarwinDevelopment() {
+    echo "";
+}
```
Let's just remove

vvbandeira commented on etc/DependencyInstaller.sh:
```diff
@@ -200,6 +200,59 @@ _installCentosRuntime() {
+    rm ./${package}.rb
+}
+
+_installDarwinRuntime() {
```
We can have just a `_installDarwin` for macOS for sake of simplicity.

vvbandeira commented on etc/DependencyInstaller.sh:
```diff
@@ -200,6 +200,59 @@ _installCentosRuntime() {
+    fi
+    brew install bison boost cmake eigen flex libomp pyqt5 python swig tcl-tk zlib
+
+    # TODO fork to openroad organization
+    brew install donn/lemon-graph/lemon-graph
```
Tom already forked, could you update to https://github.com/The-OpenROAD-Project/homebrew-lemon-graph

vvbandeira commented on etc/DependencyInstaller.sh:
```diff
@@ -242,6 +295,10 @@ case "${platform}" in
         fi
         ;;
+    "Darwin" )
+        #os=$(sw_vers -productVersion)
```
remove commented out code

vvbandeira commented on etc/DependencyInstaller.sh:
```diff
@@ -275,6 +332,23 @@ EOF
         ;;
+    "Darwin" )
+        _installDarwinRuntime
+        # TODO what is needed in dev vs. runtime?
+        #if [[ "${option}" == "dev" ]]; then
+        #    _installDarwinDev
+        #fi
+        #_installDarwinCleanUp
```
remove, see above about only having `_installDarwin`

rovinski commented:
@vvbandeira can you check on the CI? There seems to be some instability. I was failing CTS tests, then I reran, and now I'm failing read_db even though this PR doesn't touch any source code.

I will work on the change requests.

rovinski commented:
I addressed the comments and I also updated `FindTCL.cmake` so that it is no longer required to use `-DTCL_LIB_PATHS` or `-DTCL_NO_DEFAULT_PATH`. The issue was that FindTCL.cmake was only looking for the default path on Apple M1 silicon but not for Intel silicon.

vvbandeira reviewed (approved):
Thanks for the changes.
